spec/reports/middleware/cache_spec.rb

```diff
@@ -0,0 +1,151 @@
+require "faraday"
+require "time"
+require "reports/middleware/cache"
+require "reports/storage/redis"
+
+module Reports::Middleware
+  RSpec.describe Cache do
+    let(:stubs) { Faraday::Adapter::Test::Stubs.new }
+
+    let(:storage) { ::Reports::Storage::Redis.new }
+
+    let(:conn) do
+      Faraday.new do |builder|
+        builder.use Cache, ::Reports::Storage::Redis.new
+        builder.adapter :test, stubs
+      end
+    end
+
+    let(:response_array) do
+      headers = {"Cache-Control" => "max-age=300", "Date" => Time.new.httpdate}
+      [200, headers, "hello"]
+    end
+
+    after do
+      storage.flush
+    end
+
+    it "returns a previously cached response" do
+      stubs.get("http://example.test") { [200, { 'Cache-Control' => 'public max-age=60', 'Date' => Time.now.httpdate }, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(200)
+    end
+
+    %w{post patch put}.each do |http_method|
+      it "does not cache #{http_method} requests" do
+        stubs.send(http_method, "http://example.test") { [200, {'Cache-Control' => 'public'}, "hello"] }
+        conn.send(http_method, "http://example.test")
+        stubs.send(http_method, "http://example.test") { [404, {}, "not found"] }
+
+        response = conn.send(http_method, "http://example.test")
+        expect(response.status).to eql(404)
+      end
+    end
+
+    it "does not cache when the response doesn't have Cache-Controll header" do
+      stubs.get("http://example.test") { [200, {}, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "does not cache when the response Cache-Controll header has no-store value" do
+      stubs.get("http://example.test") { [200, {'Cache-Control' => 'no-store'}, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "does not use cached response when the response Cache-Controll header has no-cache value" do
+      stubs.get("http://example.test") { [200, {'Cache-Control' => 'no-store'}, "hello"] }
```

Comment:
Line is too long. [90/80]<br>Space inside { missing.<br>Space inside } missing.<br>Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.